File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/binarysearch/Problem1939.java
package com.keunwon.algorithm.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.StringTokenizer;

public class Problem1939 {
    private static final List<List<Bridge>> bridges = new ArrayList<>();
    private static int N, M;
    private static int from, to;

    private static boolean[] visited;
    private static int left = Integer.MAX_VALUE;
    private static int right = Integer.MIN_VALUE;

    public static void main(String[] args) throws IOException {
        input();
        binarySearch();
    }

    private static void binarySearch() {
        while (left <= right) {
            final var mid = (left + right) / 2;
            visited = new boolean[N + 1];

            if (bfs(mid)) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        System.out.println(right);
    }

    private static boolean bfs(int weight) {
        final var queue = new LinkedList<Integer>();

        queue.offer(from);
        visited[from] = true;

        while (!queue.isEmpty()) {
            final var now = queue.poll();

            if (now == to) { return true; }

            for (Bridge bridge : bridges.get(now)) {
                if (!visited[bridge.to] && weight <= bridge.weight) {
                    queue.offer(bridge.to);
                    visited[bridge.to] = true;
                }
            }
        }
        return false;
    }

    private static void input() throws IOException {
        final var br = new BufferedReader(new InputStreamReader(System.in));

        try (br) {
            var st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            M = Integer.parseInt(st.nextToken());

            for (int i = 0; i <= N; i++) {
                bridges.add(new ArrayList<>());
            }

            for (int i = 0; i < M; i++) {
                st = new StringTokenizer(br.readLine());
                final var landA = Integer.parseInt(st.nextToken());
                final var landB = Integer.parseInt(st.nextToken());
                final var weight = Integer.parseInt(st.nextToken());

                bridges.get(landA).add(new Bridge(landB, weight));
                bridges.get(landB).add(new Bridge(landA, weight));

                left = Math.min(left, weight);
                right = Math.max(right, weight);
            }

            st = new StringTokenizer(br.readLine());
            from = Integer.parseInt(st.nextToken());
            to = Integer.parseInt(st.nextToken());
        }
    }

    private static class Bridge {
        final int to;
        final int weight;

        Bridge(int to, int weight) {
            this.to = to;
            this.weight = weight;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem1939.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 중량제한
 * Level: 골드-3
 **/
class Problem1939 {
    private lateinit var graph: Array<MutableList<Node>>
    private lateinit var position: Pair<Int, Int>

    fun solution(n: Int, edges: Array<Triple<Int, Int, Int>>, position: Pair<Int, Int>): Int {
        this.graph = Array(n + 1) { mutableListOf() }
        this.position = position
        var left = Int.MAX_VALUE
        var right = 0

        edges.forEach { (a, b, c) ->
            graph[a].add(Node(b, c))
            graph[b].add(Node(a, c))
            left = left.coerceAtMost(c)
            right = right.coerceAtLeast(c)
        }

        while (left <= right) {
            val mid = (left + right) / 2

            if (bfs(n, mid)) left = mid + 1
            else right = mid - 1
        }
        return right
    }

    private fun bfs(n: Int, limit: Int): Boolean {
        val queue = LinkedList<Int>()
        val visited = BooleanArray(n + 1)

        with(position.first) {
            queue.offer(this)
            visited[this] = true
        }

        while (queue.isNotEmpty()) {
            val cur = queue.poll()

            if (cur == position.second) return true

            for (next in graph[cur]) {
                if (visited[next.index] || limit > next.limit) continue

                visited[next.index] = true
                queue.offer(next.index)
            }
        }
        return false
    }

    private data class Node(val index: Int, val limit: Int)
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val readIntArray = {
            br.readLine().split(" ").map { it.toInt() }.toIntArray()
        }

        val (n, m) = readIntArray()
        val edges = Array(m) {
            readIntArray().let { Triple(it[0], it[1], it[2]) }
        }
        val dest = readIntArray().let { it[0] to it[1] }

        Problem1939().solution(n, edges, dest).also { bw.write("$it") }
        bw.flush()
    }
}
